### ISSUE TITLE
Race condition when using LogWriter

### ISSUE THREAD

**Issue by cpheps** [firecracker-microvm/firecracker-go-sdk#156]
After #145 was merged there now exists a race condition when setting up a LogWriter

In `captureFifoToFile`  this goroutine

```Go
	go func() {
		defer func() {
			if err := fifoPipe.Close(); err != nil {
				logger.Warnf("Failed to close fifo pipe: %v", err)
			}

			if err := syscall.Unlink(fifoPath); err != nil {
				logger.Warnf("Failed to unlink %s: %v", fifoPath, err)
			}
		}()

		if _, err := io.Copy(w, fifoPipe); err != nil {
			logger.Warnf("io.Copy failed to copy contents of fifo pipe: %v", err)
		}
	}()
```

will not block on `io.Copy`, I think this is due to opening with `syscall.O_NONBLOCK` unsure though. This causes the defer to trigger and `syscall.Unlink` to remove the originally configured `log.fifo`.  When using jailer if this happens before the `LinkFilesHandler` handler executes you can get a missing file error when trying to hardlink the `log.fifo` shown [here](https://github.com/firecracker-microvm/firecracker-go-sdk/blob/master/jailer.go#L418-L424).

If not using jailer I think that `io.Copy` could just exit and `BootstrapLoggingHandler` could have an error if it sent it's logging setup request after the `syscall.Unlink` occurs.

**Comment by xibz** [comment]
Hey @cpheps, thank you for bringing this to our attention. Out of curiosity what is the error that is being returned from the io.Copy?

**Comment by cpheps** [comment]
That's the weird thing, there is no error. I'm guessing because `io.Copy` exits on an EOF with the `syscall.O_NONBLOCK` we are getting an early EOF with no writer to the fifo. Unsure on this just speculating.

**Comment by xibz** [comment]
Oh? That’s interesting.
What happens when you run your application?
What is the error?
Can you copy and paste the output here?

**Comment by cpheps** [comment]
Yeah the error is 
`machine link /home/cphelps/jailer/log.fifo /home/cphelps/jailer/firecracker/662/root/log.fifo: no such file or directory` that comes out of the  `LinkFilesHandler` inside jailer.go.

**Comment by xibz** [comment]
Thank you @cpheps for the information. I will go ahead and write a small test to test the behavior of nonblock and copy. Once I have more information from the test and running the jailer manually, Ill take a look at what options we have!

**Comment by xibz** [comment]
@cpheps - I am having trouble reproducing this. Here is my example

```go
package main

import (
	"context"
	"os"

	"github.com/firecracker-microvm/firecracker-go-sdk"
	"github.com/firecracker-microvm/firecracker-go-sdk/client/models"
)

func main() {
	f, err := os.OpenFile("/tmp/writer.fifo", os.O_CREATE|os.O_RDWR, 0666)
	if err != nil {
		panic(err)
	}
	defer f.Close()

	cfg := firecracker.Config{
		SocketPath:      "./api.socket",
		LogFifo:         "/tmp/log.fifo",
		LogLevel:        "Debug",
		MetricsFifo:     "/tmp/metrics.fifo",
		KernelImagePath: "hello-vmlinux.bin",
		Drives:          firecracker.NewDrivesBuilder("hello-rootfs.ext4").Build(),
		FifoLogWriter:   f,
		KernelArgs:      "console=ttyS0 reboot=k panic=1 pci=off",
		JailerCfg:       &firecracker.JailerConfig{},
	}
	cfg.MachineCfg = models.MachineConfiguration{}
	cfg.MachineCfg.VcpuCount = firecracker.Int64(1)
	cfg.MachineCfg.MemSizeMib = firecracker.Int64(512)
	cfg.MachineCfg.HtEnabled = firecracker.Bool(false)
	cfg.JailerCfg.UID = firecracker.Int(123)
	cfg.JailerCfg.GID = firecracker.Int(123)
	cfg.JailerCfg.NumaNode = firecracker.Int(0)
	cfg.JailerCfg.ID = "test"
	cfg.JailerCfg.ChrootBaseDir = "/tmp/test"
	cfg.JailerCfg.ExecFile = "/usr/local/bin/firecracker"
	cfg.JailerCfg.ChrootStrategy = firecracker.NewNaiveChrootStrategy("/tmp/test/firecracker/test", "hello-vmlinux.bin")
	cfg.JailerCfg.Stdin = os.Stdin
	cfg.JailerCfg.Stdout = os.Stdout
	cfg.JailerCfg.Stderr = os.Stderr

	m, err := firecracker.NewMachine(context.Background(), cfg)
	if err != nil {
		panic(err)
	}

	if err := m.Start(context.Background()); err != nil {
		panic(err)
	}

	if err := m.Wait(context.Background()); err != nil {
		panic(err)
	}
}
```

Can you run that example and see if that returns the same error? If it does not, can you paste some of your code to see if I can reproduce it on my end?

Also, what Go version are you using? Just so I am using the same version as you.

**Comment by cpheps** [comment]
I'll try it out now. Using go1.12.5 on linux/amd64

**Comment by cpheps** [comment]
@xibz I tried it and did not see the error, though I did notice nothing was ever written to the `writer.fifo`.  I was able to reproduce it though by vendoring the dependencies and adding a `time.Sleep(1 * time.Second)` right [here](https://github.com/firecracker-microvm/firecracker-go-sdk/blob/master/jailer.go#L413). A sleep should be fine if `io.Copy` is blocking then no weird behavior should be expected.
